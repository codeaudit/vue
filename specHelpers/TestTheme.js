import {setOpacity, toRgbaString} from '@andrew-codes/color-functions';

export default {
    _name: 'Test Theme',
    altColor: '#ffa500',
    basicColor: '#00ffff',
    basicFontFamily: '\'Proxima Nova\', \'Lucida Sans Unicode\', \'Lucida Grande\', sans-serif',
    baseIconSize: 24,
    boldFont: 600,
    borderPrimaryColor: '#000',
    darkInverseColor: '#fff',
    disabledPrimaryColor: '#808080',
    errorPrimaryColor: '#ffc0cb',
    errorSecondaryColor: '#ffc0cd',
    fieldBorderColor: '#808080',
    focusedPrimaryColor: '#00ffff',
    focusedSecondaryColor: toRgbaString(setOpacity('#00ffff', 0.5)),
    gutter: 8,
    importantColor: '#f99643',
    largeFontSize: 22,
    largeGutter: 12,
    largeLineHeight: 2.285,
    lightInverseColor: '#030303',
    mediumFontSize: 16,
    normalBackground: '#fff',
    normalLineHeight: 1.285,
    normalRadius: 3,
    requiredPrimaryColor: '#ffc0cb',
    pendingPrimaryColor: '#ffff00',
    smallFontSize: 14,
    smallGutter: 6,
    textPrimaryColor: '#030303',
    textSecondaryColor: '#030303',
    textDisabledColor: '#808080',
    xLargeFontSize: 24,
    xLargeGutter: 16,
    xMediumFontSize: 18,
    xSmallFontSize: 8,
    xxSmallGutter: 3
};
