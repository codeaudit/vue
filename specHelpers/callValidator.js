import ReactPropTypesArgs from './reactPropTypeArgs';

const callValidator = (validator, element, propName = '', componentName = '') => validator(element.props, propName, componentName, ...ReactPropTypesArgs);

export const validatorPasses = (validator, element, propName) => callValidator(validator, element, propName) === null;
export const validatorFails = (validator, element, propName) => callValidator(validator, element, propName) instanceof Error;
