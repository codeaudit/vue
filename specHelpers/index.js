import React from 'react';
import ThemeProvider, {getTheme} from './../src/Theme';

export const applyTheme = (ComponentUnderTest, theme = getTheme()) => <ThemeProvider theme={theme}>{ComponentUnderTest}</ThemeProvider>;
