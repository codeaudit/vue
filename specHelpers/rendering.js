import toJson from 'enzyme-to-json';
import {mount, shallow} from 'enzyme';
import React from 'react';
import theme from './TestTheme';

export const getRender = (Component) => (props = {}, context = {theme}) => shallow(<Component {...props} />, {context});
export const getMount = (Component) => (props = {}, context = {theme}) => mount(<Component {...props} />, {context});
export const snapshot = (renderedComponent) => toJson(renderedComponent);
