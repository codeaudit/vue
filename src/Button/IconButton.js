import React, {PropTypes} from 'react';
import ui from 'redux-ui';
import {normal} from './Sizes';
import Radium from './../utilities/Radium';
import transparent from './../utilities/Transparent';
import {create as createTransition} from '../utilities/Transitions';
import {createConditionalEventHandler, createEventHandlerIgnoringEventData} from './../utilities/component';

const getStyles = (props, theme) => ({
    root: {
        backgroundColor: (Boolean(props.disabled) || (!props.hovered && !props.ui.hovered)) ? props.backgroundColor : props.hoverBackgroundColor,
        border: (props.disabled && Boolean(props.border)) ? `1px solid ${theme.disabledPrimaryColor}` : Boolean(props.border) ? props.border : `1px solid ${transparent}`,
        borderRadius: props.circle ? '50%' : '0px',
        cursor: props.disabled ? 'not-allowed' : 'pointer',
        display: 'inline-block',
        lineHeight: 0.6,
        transition: props.transition,
    },
});

const IconButton = (props, context) => {
    const {
        color,
        disabled,
        hoverColor,
        hovered,
        ui,
        onClick,
        size,
        updateUI,
    } = props;
    const {
        baseIconSize,
        disabledPrimaryColor,
        smallGutter,
    } = context.theme;

    const width = baseIconSize * size;
    const iconColor = disabled ? disabledPrimaryColor : color;
    const iconHoverColor = disabled ? disabledPrimaryColor : hoverColor;
    const handleClick = createConditionalEventHandler(!disabled)(onClick);
    const handleMouseEnter = createEventHandlerIgnoringEventData(updateUI, 'hovered', true);
    const handleMouseLeave = createEventHandlerIgnoringEventData(updateUI, 'hovered', false);
    const styles = getStyles(props, context.theme);

    return (
        <div
            style={styles.root}
            onClick={handleClick}
            onMouseEnter={handleMouseEnter}
            onMouseLeave={handleMouseLeave}
        >
            {React.createElement(props.icon, {
                color: iconColor,
                hoverColor: iconHoverColor,
                hovered: hovered || ui.hovered,
                padding: smallGutter,
                width,
            })}
        </div>
    );
};
IconButton.propTypes = {
    /**
     * Background color of button
     */
    backgroundColor: PropTypes.string,
    /**
     * Border of icon button
     */
    border: PropTypes.string,
    /**
     * Will render a circular button when true
     */
    circle: PropTypes.bool,
    /**
     * Color of the SvgIcon within button
     */
    color: PropTypes.string,
    /**
     * Disables the button from responding to event handlers
     */
    disabled: PropTypes.bool,
    /**
     * Background color when hovered
     */
    hoverBackgroundColor: PropTypes.string,
    /**
     * Hover color of SvgIcon
     */
    hoverColor: PropTypes.string,
    /**
     * Indicates the IconButton is in a hovered state when true
     */
    hovered: PropTypes.bool,
    /**
     * Icon to render within button
     */
    icon: PropTypes.func.isRequired,
    /**
     * Click event handler; fired once a button is clicked
     */
    onClick: PropTypes.func,
    /**
     * Numeric value used as a multiplier to the button's size; 0.75, 1, and 1.5 as examples
     */
    size: PropTypes.number,
    transition: PropTypes.string,
    /**
     * Managed UI state props; can be overridden
     */
    ui: PropTypes.object,
};
IconButton.defaultProps = {
    backgroundColor: transparent,
    circle: false,
    disabled: false,
    onClick: () => {
    },
    size: normal,
    transition: createTransition('0.25s', 'all', '0ms', 'linear'),
};
IconButton.contextTypes = {
    theme: PropTypes.object.isRequired,
};
IconButton.displayName = 'IconButton';
export default Radium(ui({
    state: {
        hovered: false,
    }
})(IconButton));
