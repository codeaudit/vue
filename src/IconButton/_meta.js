import examples from './examples';

export default {
    name: 'Icon Button',
    keywords: ['basic', 'important', 'alt', 'special', 'icon', 'svg'],
    'see also': ['Button', 'Icons', 'SvgIcon'],
    readme: require('./README.md'),
    status: 'experimental',
    componentsSources: [
        {
            name: 'IconButton',
            code: require('!raw!./IconButton')
        }
    ],
    examples
};
