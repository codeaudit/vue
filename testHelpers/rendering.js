import toJson from 'enzyme-to-json';
import React, {PropTypes} from 'react';
import {mount, render, shallow} from 'enzyme';
import theme from './../packages/themes/src/V1Theme';

const ctx = {
    theme,
};

export const getShallow = (Component) => (props = {}, context = {}) => shallow(<Component {...props} />, {
    childContextTypes: {
        theme: PropTypes.object,
    },
    context: {
        ...ctx,
        ...context,
    },
});

export const getRender = (Component) => (props = {}, context = {}) => render(<Component {...props} />, {
    childContextTypes: {
        theme: PropTypes.object,
    },
    context: {
        ...ctx,
        ...context,
    },
});

export const getMount = (Component) => (props = {}, options = {}, context = {}) => mount(<Component {...props} />, {
    childContextTypes: {
        theme: PropTypes.object,
    },
    context: {
        ...ctx,
        ...context,
    },
    ...options,
});

const replaceUIKey = (componentName) => {
    let index = 0;
    return (json) => {
        if (!json.props || !json.props.uiKey) {
            return json;
        }
        const uiKey = `${componentName}-${index}`;
        index += 1;
        return Object.assign({}, json, {
            uiKey,
            uiPath: [uiKey],
        });
    };
};
const replaceUiKeyRecursively = (json) => {
    if (!json.type) {
        return json;
    }
    const replaceUIKeyInParent = replaceUIKey(json.type);
    const output = replaceUIKeyInParent(json);
    if (!output.children) {
        return json;
    }
    output.children = json.children.map((child, index) => replaceUiKeyRecursively(child));
    return output;
};

export const snapshot = (renderedComponent) => {
    return toJson(renderedComponent);
};

export const reset = (component) => () => {
    if (component) {
        component.unmount();
    }
    document.body.innerHTML = '';
};
