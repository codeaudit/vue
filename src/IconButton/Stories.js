import React from 'react';
import {action, storiesOf} from '@kadira/storybook';
import CloseIcon from './../Icons/CloseIcon';
import IconButton from './';

storiesOf('IconButton')
    .addWithInfo('hover colors',
        '',
        () => (
            <IconButton
                icon={CloseIcon}
                backgroundColor="white"
                color="blue"
                hoverBackgroundColor="blue"
                hoverColor="white"
                onClick={action('clicked')}
            />
        )
    )
    .addWithInfo('circular',
        '',
        () => (
            <IconButton
                backgroundColor="blue"
                circle
                color="white"
                hoverBackgroundColor="white"
                hoverColor="blue"
                icon={CloseIcon}
                size={ButtonSizes.small}
                onClick={action('clicked')}
            />
        )
    )
    .addWithInfo('border',
        '',
        () => (
            <div>
                <IconButton
                    icon={CloseIcon}
                    backgroundColor="white"
                    border="10px solid green"
                    color="blue"
                    hoverBackgroundColor="blue"
                    hoverColor="white"
                    onClick={action('clicked')}
                />
                <IconButton
                    backgroundColor="white"
                    border="10px solid green"
                    circle
                    color="blue"
                    hoverBackgroundColor="blue"
                    hoverColor="white"
                    icon={CloseIcon}
                    size={ButtonSizes.small}
                    onClick={action('clicked')}
                />
            </div>
        )
    )
    .addWithInfo('disabled',
        '',
        () => (
            <div>
                <div>
                    <h2>With Background colors</h2>
                    <IconButton
                        backgroundColor="white"
                        disabled
                        hoverBackgroundColor="blue"
                        hoverColor="white"
                        icon={CloseIcon}
                        onClick={action('clicked')}
                    />
                    <IconButton
                        backgroundColor="white"
                        border="blue"
                        disabled
                        hoverBackgroundColor="blue"
                        hoverColor="white"
                        icon={CloseIcon}
                        onClick={action('clicked')}
                    />
                    <IconButton
                        backgroundColor="white"
                        circle
                        disabled
                        hoverBackgroundColor="blue"
                        hoverColor="white"
                        icon={CloseIcon}
                        onClick={action('clicked')}
                    />
                    <IconButton
                        backgroundColor="white"
                        border="blue"
                        circle
                        disabled
                        hoverBackgroundColor="blue"
                        hoverColor="white"
                        icon={CloseIcon}
                        onClick={action('clicked')}
                    />
                </div>
                <div>
                    <h2>Without Background colors</h2>
                    <IconButton
                        disabled
                        hoverBackgroundColor="blue"
                        hoverColor="white"
                        icon={CloseIcon}
                        onClick={action('clicked')}
                    />
                    <IconButton
                        border="blue"
                        disabled
                        hoverBackgroundColor="blue"
                        hoverColor="white"
                        icon={CloseIcon}
                        onClick={action('clicked')}
                    />
                    <IconButton
                        circle
                        disabled
                        hoverBackgroundColor="blue"
                        hoverColor="white"
                        icon={CloseIcon}
                        onClick={action('clicked')}
                    />
                    <IconButton
                        border="blue"
                        circle
                        disabled
                        hoverBackgroundColor="blue"
                        hoverColor="white"
                        icon={CloseIcon}
                        onClick={action('clicked')}
                    />
                </div>
            </div>
        )
    )
    .addWithInfo('resizing',
        '',
        () => (
            <div>
                <IconButton disabled icon={CloseIcon} size={ButtonSizes.small} onClick={action('clicked')} />
                <IconButton disabled icon={CloseIcon} size={ButtonSizes.normal} onClick={action('clicked')} />
                <IconButton disabled icon={CloseIcon} size={ButtonSizes.large} onClick={action('clicked')} />
            </div>
        )
    );
