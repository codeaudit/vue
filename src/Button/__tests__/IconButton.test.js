import IconButton from './../IconButton';
import CloseIcon from './../../Icons/CloseIcon';
import {getRender, snapshot} from './../../../specHelpers/rendering';

const renderIconButton = getRender(IconButton);

test('icon rendering', () => {
    const iconButton = renderIconButton({
        icon: CloseIcon,
    });
    expect(snapshot(iconButton)).toMatchSnapshot();
});

test('circular icon', () => {
    const iconButton = renderIconButton({
        circle: true,
        icon: CloseIcon,
    });
    expect(snapshot(iconButton)).toMatchSnapshot();
});

test('it can be resized', () => {
    const button = renderIconButton({
        icon: CloseIcon,
        size: 0.75
    });
    expect(snapshot(button)).toMatchSnapshot();
});

test('it can be clicked', () => {
    const onClick = jest.fn();
    const iconButton = renderIconButton({
        icon: CloseIcon,
        onClick,
    });
    simulateClick(iconButton);
    expect(onClick).toHaveBeenCalledTimes(1);
});

test('it can be disabled', () => {
    const onClick = jest.fn();
    const iconButton = renderIconButton({
        disabled: true,
        icon: CloseIcon,
        onClick,
    });
    simulateClick(iconButton);
    expect(onClick).toHaveBeenCalledTimes(0);
    expect(snapshot(iconButton)).toMatchSnapshot();
});

function simulateClick(wrapper) {
    wrapper.simulate('click');
}
