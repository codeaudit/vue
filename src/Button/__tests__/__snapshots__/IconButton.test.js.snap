exports[`test circular icon 1`] = `
<div
  data-radium={true}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0)",
      "border": "1px solid rgba(0,0,0,0)",
      "borderRadius": "50%",
      "cursor": "pointer",
      "display": "inline-block",
      "lineHeight": 0.6,
      "transition": "all 0.25s linear 0ms",
    }
  }>
  <CloseIcon
    padding={6}
    width={24} />
</div>
`;

exports[`test icon rendering 1`] = `
<div
  data-radium={true}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0)",
      "border": "1px solid rgba(0,0,0,0)",
      "borderRadius": "0px",
      "cursor": "pointer",
      "display": "inline-block",
      "lineHeight": 0.6,
      "transition": "all 0.25s linear 0ms",
    }
  }>
  <CloseIcon
    padding={6}
    width={24} />
</div>
`;

exports[`test it can be disabled 1`] = `
<div
  data-radium={true}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0)",
      "border": "1px solid rgba(0,0,0,0)",
      "borderRadius": "0px",
      "cursor": "not-allowed",
      "display": "inline-block",
      "lineHeight": 0.6,
      "transition": "all 0.25s linear 0ms",
    }
  }>
  <CloseIcon
    color="#808080"
    hoverColor="#808080"
    padding={6}
    width={24} />
</div>
`;

exports[`test it can be resized 1`] = `
<div
  data-radium={true}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0)",
      "border": "1px solid rgba(0,0,0,0)",
      "borderRadius": "0px",
      "cursor": "pointer",
      "display": "inline-block",
      "lineHeight": 0.6,
      "transition": "all 0.25s linear 0ms",
    }
  }>
  <CloseIcon
    padding={6}
    width={18} />
</div>
`;
